src/utils/recursive.test.ts

```diff
@@ -0,0 +1,127 @@
+import { generateId } from './generateId'
+import { duplicateComponent, deleteComponent } from './recursive'
+
+jest.mock('./generateId')
+
+const mockGenerateId: jest.Mock<typeof generateId> = generateId as any
+
+describe('recursive functions', () => {
+  let id = 1
+  let mockFn: typeof mockGenerateId
+
+  beforeAll(() => {
+    // @ts-ignore
+    mockFn = mockGenerateId.mockImplementation(() => {
+      return `comp-${id++}`
+    })
+  })
+
+  afterEach(() => {
+    id = 1
+  })
+
+  afterAll(() => {
+    mockFn.mockClear()
+  })
+
+  it('should duplicate a Box component containing an Avatar', () => {
+    const initialComponents: IComponents = {}
+
+    const first = generateId()
+    const second = generateId()
+
+    initialComponents.root = {
+      id: 'root',
+      type: 'Box',
+      parent: 'root',
+      props: {},
+      children: [first],
+    }
+
+    initialComponents[first] = {
+      id: first,
+      type: 'Box',
+      parent: 'root',
+      props: {},
+      children: [second],
+    }
+
+    initialComponents[second] = {
+      id: second,
+      type: 'Avatar',
+      parent: first,
+      props: {},
+      children: [],
+    }
+
+    const { clonedComponents } = duplicateComponent(
+      initialComponents[first],
+      initialComponents,
+    )
+
+    expect(Object.keys(clonedComponents).length).toEqual(2)
+
+    const finalTree = {
```

Comment:
This is testing the implementation that we know we are doing in redux. We can remove this test and test that the cloned components have new instance.